### PR TITLE
BUG: iterating on a subset of columns in a GroupBy object (#44821)

### LINKED ISSUE #44821: BUG: Inconsistent behaviour in DataFrameGroupBy when selecting a subset of columns
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame({k: range(10) for k in "ABC"})

# Group by and select columns
dg = df.groupby(df.A < 4)[["A", "B"]]


# Apply on GroupBy
print(dg.apply(lambda x: [x.shape, x.columns]))

# A
# False    [(6, 2), [A, B]]
# True     [(4, 2), [A, B]]
# dtype: object


# Iterate on GroupBy
for ix, dg_ix in dg:
    print(ix, dg_ix.shape, dg_ix.columns)

# False (6, 3) Index(['A', 'B', 'C'], dtype='object')
# True (4, 3) Index(['A', 'B', 'C'], dtype='object')
```

### Issue Description

A subset of columns is not (always) selected with `DataFrameGroupBy.__getitem__`. Selecting a subset of column works for `DataFrameGroupBy.apply` but not for `DataFrameGroupBy.__iter__`.

### Expected Behavior

I expected `dg_ix` to contain only `A` and `B` as columns.

```python
# Iterate on GroupBy
for ix, dg_ix in dg:
    print(ix, dg_ix.shape, dg_ix.columns)

# False (6, 2) Index(['A', 'B'], dtype='object')
# True (4, 2) Index(['A', 'B'], dtype='object')
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.14.18-300.fc35.x86_64
Version          : #1 SMP Fri Nov 12 16:43:17 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 59.1.0
Cython           : None
pytest           : 5.4.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 7.29.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.2
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : 0.20.1
xlrd             : None
xlwt             : None
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by maurosilber**
Fixes issue #44821.

When trying to iterate on a subset of columns in a GroupBy object,
it returned all columns, instead of the selected subset.

GroupBy.__iter__ used self.obj instead of self._selected_obj (see
PR #6570).

- [x] closes #44821
- [x] 1 test added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by phofl — CHANGES_REQUESTED** [review]
Please run the tests locally first. This broke lots of tests

**Comment by maurosilber** [comment]
Hi.

Only one test is broken, which I mentioned it here: https://github.com/pandas-dev/pandas/issues/44821#issuecomment-997037279

As mentioned there, I'm not sure how to continue, as users may rely on current behavior.

By the way, should I could continue the discussion in this PR?

**Comment by maurosilber** [comment]
Well, I decided to fix that test. While this PR changes `GroupBy.__iter__`, it didn't actually broke any test that tested its behavior. The test that was previously failing concerned the behaviour of `GroupBy.rolling`, and the fix was in the "manual computation" to compare against.

So, although this PR might break somebody's code, this `GroupBy` behavior wasn't "promised" anywhere, and it could have been changed as part of #6570.

**Review by jreback — CHANGES_REQUESTED** [review]
ok this is reasonable, can you add a whatsnew note in 1.4, groupby bug fix section. you may want to give a mini-example of what the change is

**Comment by jreback** [comment]
cc @rhshadrach for comments

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -797,6 +797,7 @@ Groupby/resample/rolling
 - Bug in :meth:`GroupBy.nth` failing on ``axis=1`` (:issue:`43926`)
 - Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` not respecting right bound on centered datetime-like windows, if the index contain duplicates (:issue:`3944`)
 - Bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` when using a :class:`pandas.api.indexers.BaseIndexer` subclass that returned unequal start and end arrays would segfault instead of raising a ``ValueError`` (:issue:`44470`)
+- Fixed bug in :meth:`GroupBy.__iter__` after selecting a subset of columns in a :class:`GroupBy` object, which returned all columns instead of the chosen subset (:issue:`#44821`).
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.4.0.rst:800]: nit: No period after the issue number.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Small nitpick, otherwise I think this is good.

There is an oddity with _selected_obj, but I think this change is definitely an improvement over current behavior and it is really independent, so still +1. Once you subset the columns of a groupby (e.g. `['b']` or `[['b']]`), the selected_obj will not include the grouper. However, before you do this, it will unless the context is active. I.e.

```
df = pd.DataFrame({'a': [1, 2], 'b': [3, 4], 'c': [5, 6]})
gb = df.groupby('a')
print(list(gb.grouper.get_iterator(gb._selected_obj, axis=gb.axis)))
print()
with gb._group_selection_context():
    print(list(gb.grouper.get_iterator(gb._selected_obj, axis=gb.axis)))
    print()
selected_gb = gb[['b']]
print(list(selected_gb.grouper.get_iterator(selected_gb._selected_obj, axis=gb.axis)))
```

produces

```
[(1,    a  b  c
0  1  3  5), (2,    a  b  c
1  2  4  6)]

[(1,    b  c
0  3  5), (2,    b  c
1  4  6)]

[(1,    b
0  3), (2,    b
1  4)]
```

In the first print, the grouper 'a' is present. In the 2nd and 3rd it isn't.

File: doc/source/whatsnew/v1.4.0.rst (hunk 1 above)
Comment (maurosilber) [reply doc/source/whatsnew/v1.4.0.rst]: Done.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @maurosilber

**Pull request merged**